vsintegration/src/FSharp.Editor/Formatting/IndentationService.fs

```diff
@@ -12,35 +12,80 @@ open Microsoft.CodeAnalysis.Formatting
 open Microsoft.CodeAnalysis.Host.Mef
 open Microsoft.CodeAnalysis.Text
 
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
 [<Shared>]
 [<ExportLanguageService(typeof<ISynchronousIndentationService>, FSharpConstants.FSharpLanguageName)>]
-type internal FSharpIndentationService() =
+type internal FSharpIndentationService
+    [<ImportingConstructor>]
+    (projectInfoManager: ProjectInfoManager) =
 
-    static member GetDesiredIndentation(sourceText: SourceText, lineNumber: int, tabSize: int): Option<int> =        
+    static member GetDesiredIndentation(documentId: DocumentId, sourceText: SourceText, filePath: string, lineNumber: int, tabSize: int, optionsOpt: FSharpProjectOptions option): Option<int> =
         // Match indentation with previous line
         let rec tryFindPreviousNonEmptyLine l =
             if l <= 0 then None
             else
                 let previousLine = sourceText.Lines.[l - 1]
-                if not (String.IsNullOrEmpty(previousLine.ToString())) then                    
+                if not (String.IsNullOrEmpty(previousLine.ToString())) then
                     Some previousLine
                 else
                     tryFindPreviousNonEmptyLine (l - 1)
-        // No indentation on the first line of a document
-        if lineNumber = 0 then None
-        else
-            match tryFindPreviousNonEmptyLine lineNumber with
-            | None -> Some 0
-            | Some previousLine ->          
-                let rec loop column spaces =
-                    if previousLine.Start + column >= previousLine.End then
-                        spaces
-                    else 
-                        match previousLine.Text.[previousLine.Start + column] with
-                        | ' ' -> loop (column + 1) (spaces + 1)
-                        | '\t' -> loop (column + 1) (((spaces / tabSize) + 1) * tabSize)
-                        | _ -> spaces
-                Some (loop 0 0)
+
+        let rec tryFindLastNoneEmptyToken (line: TextLine) = maybe {
```

Comment:
Maybe rename this from `tryFindLastNoneEmptyToken` to `tryFindLastNonWhitespaceToken`